x-pack/plugins/ml/public/application/services/dashboard_service.ts

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'kibana/public';
+import { htmlIdGenerator } from '@elastic/eui';
+import { useMemo } from 'react';
+import {
+  DASHBOARD_APP_URL_GENERATOR,
+  DashboardUrlGenerator,
+  SavedDashboardPanel,
+  SavedObjectDashboard,
+} from '../../../../../../src/plugins/dashboard/public';
+import { useMlKibana } from '../contexts/kibana';
+import { ViewMode } from '../../../../../../src/plugins/embeddable/public';
+
+export type DashboardService = ReturnType<typeof dashboardServiceProvider>;
+
+export function dashboardServiceProvider(
+  savedObjectClient: SavedObjectsClientContract,
+  kibanaVersion: string,
+  dashboardUrlGenerator: DashboardUrlGenerator
+) {
+  const generateId = htmlIdGenerator();
+  const DEFAULT_PANEL_WIDTH = 24;
+  const DEFAULT_PANEL_HEIGHT = 15;
+
+  return {
+    /**
+     * Fetches dashboards
+     */
+    async fetchDashboards(query?: string) {
+      return await savedObjectClient.find<SavedObjectDashboard>({
+        type: 'dashboard',
+        perPage: 10,
+        search: query ? `${query}*` : '',
+        searchFields: ['title^3', 'description'],
+      });
+    },
+    /**
+     * Resolves the last positioned panel from the collection.
+     */
+    getLastPanel(panels: SavedDashboardPanel[]): SavedDashboardPanel | null {
+      return panels.length > 0
+        ? panels.reduce((prev, current) =>
+            prev.gridData.y >= current.gridData.y
+              ? prev.gridData.y === current.gridData.y
+                ? prev.gridData.x > current.gridData.x
+                  ? prev
+                  : current
+                : prev
+              : current
+          )
+        : null;
+    },
+    /**
+     * Attaches embeddable panels to the dashboard
+     */
+    async attachPanels(
```

Comment:
You could potentially simplify and standardize the placement of the new panel by using the `findTopLeftMostOpenSpace` method in `src/plugins/dashboard/public/application/embeddable/panel/dashboard_panel_placement.ts`